File: packages/engine/tests/browser/drawing-expansion.mjs
export async function verifyDrawing(engine,step,load){
 const checks=[];const assert=(ok,msg)=>{if(!ok)throw Error(msg);};
 const capture=async()=>{const bitmap=await createImageBitmap(await engine.capture_frame());const canvas=document.createElement('canvas');canvas.width=bitmap.width;canvas.height=bitmap.height;const ctx=canvas.getContext('2d');ctx.drawImage(bitmap,0,0);bitmap.close();return(x,y)=>[...ctx.getImageData(x,y,1,1).data];};
 const near=(actual,expected,label)=>assert(expected.every((v,i)=>Math.abs(actual[i]-v)<4),`${label}: ${actual} != ${expected}`);
 engine.set_asset_texture('drawing-test',1,1,new Uint8Array([128,64,32,128]));
 const drawing=`draw::background(color: $COLOR_BLACK)
 transform::push()
 transform::translate(x: 40,y: 40)
 draw::rect(width: 80,height: 80,corner_radius: 10,color: $COLOR_CORAL)
 transform::pop()
 draw::image(asset: asset::bitmap(id: "drawing-test"),x: 140,y: 40,width: 80,height: 80)
 draw::rect(x: 240,y: 40,width: 80,height: 80,color: color::mix(a: $COLOR_BLACK,b: $COLOR_WHITE,amount: 0.5))
 draw::circle(x: 380,y: 80,radius: 40,gradient: color::radial_gradient(x: 380,y: 80,radius: 40,color_stops: [[0,$COLOR_WHITE],[1,$COLOR_BLACK]]))
 draw::polyline(points: [[40,180],[100,140],[160,180]],stroke_width: 8,color: $COLOR_WHITE,line_join: "round")
 draw::bezier(points: [[200,180],[240,120],[280,240],[320,180]],stroke_width: 5,color: $COLOR_WHITE)
 draw::arc(x: 380,y: 180,radius: 30,sweep_deg: 270,stroke_width: 5,color: $COLOR_WHITE)
 draw::text(content: "Drawing expansion",x: 40,y: 250,size: 24,color: $COLOR_WHITE)`;
 load(`render {${drawing}}`);step();let p=await capture();near(p(80,80),[255,127,80,255],'transformed rounded rectangle');near(p(160,80),[64,32,16,255],'straight alpha image');near(p(280,80),[188,188,188,255],'linear light mixing');assert(p(380,80)[0]>245,'radial centre');checks.push('2D transforms, rounded shapes, images, gradients, curves and colour mixing');
 load(`context 3d render {draw::cube() gfx::overlay(enabled: true) ${drawing}}`);step();p=await capture();near(p(80,80),[255,127,80,255],'overlay rectangle');near(p(160,80),[64,32,16,255],'overlay image');checks.push('3D overlay shares 2D drawing and capture dimensions');
 load(`context 3d render {draw::background(color: $COLOR_BLACK) camera::orthographic(height: 4) draw::rect(x: -1,y: -1,width: 2,height: 2,color: $COLOR_CORAL) draw::line(x1: -1,y1: 0,z1: 1,x2: 1,y2: 0,z2: 1,stroke_width: 0.1,color: $COLOR_WHITE)}`);step();p=await capture();near(p(640,360),[255,255,255,255],'thick world line');near(p(640,420),[255,127,80,255],'planar world rectangle');checks.push('3D planar fills and world-unit thick lines');
 load(`prop frames=[] on_frame {frames=[$FRAME_INDEX,$TIME_HOUR,$TIME_MINUTE,$TIME_DAY,$TIME_MONTH,$TIME_YEAR,$DELTA_SEC]} render {draw::background(color: $COLOR_BLACK)}`);step();const before=engine.get_properties();await capture();assert(engine.get_properties()===before,'capture advanced snapshot');step();assert(engine.get_properties()!==before,'frame index did not advance');checks.push('frame/calendar snapshot remains stable through capture');
 load(`context 3d render {camera::orthographic(height: 4) draw::point_cloud(count: 1,x: math::noise(x: $POINT_INDEX,seed: 7)*0,y: math::random(seed: 7,index: $POINT_INDEX)*0,z: 0,size: 30,color: color::mix(a: $COLOR_BLACK,b: $COLOR_WHITE,amount: 0.5))}`);step();p=await capture();near(p(640,360),[188,188,188,255],'GPU helpers');checks.push('GPU noise/random/colour helpers compile and render');
 // GPU-dependent inputs prevent constant folding and exercise hash/noise parity.
 const hash=n=>{n=(n^(n>>>16))>>>0;n=Math.imul(n,0x7feb352d)>>>0;n=(n^(n>>>15))>>>0;n=Math.imul(n,0x846ca68b)>>>0;return(n^(n>>>16))>>>0;};
 const random=(seed,index)=>(hash(index^hash(seed))>>>8)/16777216;
 load(`context 3d render {camera::orthographic(height: 4) draw::point_cloud(count: 4,x: $POINT_INDEX - 1.5,y: 0,z: 0,size: 30,color: color::rgb(r: math::random(seed: 7,index: $POINT_INDEX),g: math::noise(x: $POINT_INDEX,y: 0,z: 0,seed: 7)))}`);step();p=await capture();
 for(let i=0;i<4;i++){const noise=(hash(hash(hash(7^i)))>>>8)/16777216;near(p(640+(i-1.5)*180,360),[Math.round(random(7,i)*255),Math.round(noise*255),0,255],`GPU sample ${i}`);}
 checks.push('CPU/GPU random and integer-lattice noise test vectors');
 // A connected translucent corner should be no darker/brighter than its segments.
 load(`context 3d render {draw::background(color: $COLOR_BLACK) camera::orthographic(height: 4) draw::polyline(points: [[-1,0,0],[0,0,0],[0,1,0]],stroke_width: 0.4,line_join: "round",color: color::rgb(r: 1,g: 1,b: 1,a: 0.5))}`);step();p=await capture();near(p(630,370),[128,128,128,255],'translucent join');
 checks.push('continuous translucent join');
 // Verify that flat shading changes the normals used by a smooth mesh.
 const lit=mode=>`context 3d render {draw::background(color: $COLOR_BLACK) light::directional(x: -1,y: -1,z: -1) draw::sphere(radius: 2,resolution: 8,shading: "${mode}")}`;
 load(lit('flat'));step();const flat=await capture();load(lit('lambert'));step();const smooth=await capture();let differs=false;for(let x=500;x<780;x+=20){if(Math.abs(flat(x,330)[0]-smooth(x,330)[0])>5)differs=true;}assert(differs,'flat and Lambert shading were identical');checks.push('flat face normals differ from Lambert vertex normals');
 const sprite=roll=>`context 3d render {draw::background(color: $COLOR_BLACK) camera::orthographic(height: 4) draw::sprite(width: 2,height: 0.4,rotation_z_deg: ${roll},color: $COLOR_WHITE)}`;
 load(sprite(0));step();p=await capture();near(p(780,360),[255,255,255,255],'unrolled sprite');load(sprite(90));step();p=await capture();near(p(780,360),[0,0,0,255],'rolled sprite');near(p(640,220),[255,255,255,255],'visible sprite roll');checks.push('billboard Z roll');
 engine.set_asset_texture('081c271e-67d9-4a75-aff2-5a1bb88e4daa',1,1,new Uint8Array([255,255,255,255]));
 for(const name of ['01-calendar-clock','02-transform-mandala','03-connected-waveform','04-image-halo','05-world-ribbons','06-seeded-fields']){const source=await(await fetch(`../../examples/drawing/${name}.viscript`)).text();assert(!engine.validate_script(source),`${name}: validator`);load(source);for(let i=0;i<3;i++)step();await capture();checks.push(`live/capture ${name}`);}
 return {passed:true,checks};
}

File: packages/engine/tests/drawing_expansion.rs
use std::cell::RefCell;
use visamp_2::{
    interpreter::{interpret_event_block, interpret_render_block, Runtime, Target},
    model::{BlockType, Model, Value},
    parser::build_ast,
    scene::Scene,
};
fn scene(source: &str) -> Result<Scene, String> {
    let script = build_ast(source)?;
    let mut m = Model::from_script(&script);
    let rt = Runtime::new();
    let scene = RefCell::new(Scene::default());
    for b in &m.blocks {
        if b.block_type == BlockType::Render {
            interpret_render_block(
                b,
                &mut m.decels,
                Target::scene(&scene, &RefCell::new(Vec::new())),
                &rt,
                &m.functions,
            )?;
        }
    }
    Ok(scene.into_inner())
}
fn value(expr: &str) -> Result<Value, String> {
    let script = build_ast(&format!(
        "prop result=0 on_frame {{ result={expr} }} render {{}}"
    ))?;
    let mut m = Model::from_script(&script);
    interpret_event_block(&m.blocks[0], &mut m.decels, &Runtime::new(), &m.functions)?;
    Ok(m.decels.get("result").unwrap().clone())
}
#[test]
fn helpers_have_defined_edges() {
    assert_eq!(
        value("array::length(value: [1,2,3])").unwrap(),
        Value::Integer(3)
    );
    assert_eq!(
        value("array::length(value: audio::detect::get_frequency())").unwrap(),
        Value::Integer(1024)
    );
    assert_eq!(
        value("math::lerp(a: 2,b: 4,amount: 2)").unwrap(),
        Value::Float(6.0)
    );
    assert_eq!(
        value("math::wrap(value: -1,min: 0,max: 5)").unwrap(),
        Value::Float(4.0)
    );
    assert_eq!(
        value("math::smoothstep(value: 0.5,min: 0,max: 1)").unwrap(),
        Value::Float(0.5)
    );
    assert_eq!(value("math::map(value: 3,input_min: 2,input_max: 0,output_min: 0,output_max: 10,clamp: true)").unwrap(),Value::Float(0.0));
    for expr in [
        "math::map(value: 1,input_min: 0,input_max: 0,output_min: 0,output_max: 1)",
        "math::wrap(value: 1,min: 2,max: 0)",
        "math::random(seed: -0.5,index: 0)",
        "math::random(seed: 0,index: 16777216)",
        "math::noise(x: 1000001)",
        "color::radial_gradient(x: 0,y: 0,radius: 0,color_stops: [])",
    ] {
        assert!(value(expr).is_err(), "{expr}");
    }
    let Value::Color(c) = value("color::mix(a: $COLOR_BLACK,b: $COLOR_WHITE,amount: 0.5)").unwrap()
    else {
        panic!()
    };
    assert!((c.r - 0.735356983).abs() < 1e-8);
}
#[test]
fn seeded_vectors_are_stable() {
    use visamp_2::creative_math::{noise, random};
    assert_eq!(random(0, 0), 0.0);
    assert_eq!(random(1, 2), random(1, 2));
    assert_ne!(random(1, 2), random(2, 1));
    assert_eq!(noise(0.0, 0.0, 0.0, 0), 0.0);
    assert!((noise(0.5, -0.5, 0.25, 7) - noise(0.50001, -0.5, 0.25, 7)).abs() < 0.0001);
}
#[test]
fn planar_shapes_and_connected_paths_record_meshes() {
    let s=scene(r#"context 3d render {
 draw::background(color: $COLOR_CORAL)
 draw::polygon(points: [[0,0],[2,0],[2,2],[1,1],[0,2]])
 draw::rect(x: -2,width: 1,height: 1,corner_radius: 0.2)
 draw::circle(radius: 1)
 draw::ellipse(radius_x: 2,radius_y: 1)
 draw::polyline(points: [[0,0,0],[1,0,0],[1,1,1]],stroke_width: 0.1,line_join: "round",line_cap: "round")
 draw::bezier(points: [[0,0,0],[1,2,0],[2,-1,0],[3,0,0]],stroke_width: 0.1)
 draw::arc(radius: 2,sweep_deg: -90,stroke_width: 0.1)
 gfx::overlay(enabled: true)
 transform::push()
 draw::text(content: "screen")
 transform::pop()
 }"#).unwrap();
    assert_eq!(s.commands.len(), 7);
    assert_eq!(s.meshes[0].indices.len(), 9);
    assert_eq!(s.overlay_calls.len(), 3);
    assert!(s.gfx.clear.is_some());
}
#[test]
fn invalid_shapes_and_combinations_fail() {
    for source in [
 "context 3d render {draw::text(content: \"bad\")}",
 "render {draw::arc(radius: 1)}",
 "render {draw::polyline(points: [],line_join: \"miter\")}",
 "context 3d render {draw::circle(gradient: color::linear_gradient())}",
 "context 3d render {draw::sprite(rotation_x_deg: 20)}",
 "context 3d render {gfx::overlay(enabled: true) transform::rotate_x(deg: 1)}",
 "context 3d render {draw::background(texture: asset::bitmap(id: \"x\"))}",
 "render {transform::translate(z: 1)}",
 "context 3d render {draw::polygon(points: [[0,0],[1,1],[0,1],[1,0]])}",
 "context 3d render {draw::polyline(points: [[0,0],[1,1]])}",
 "context 3d render {gfx::overlay(enabled: true) draw::text(content: \"a\") gfx::overlay(enabled: false) draw::cube()}"
 ]{assert!(scene(source).is_err(),"{source}");}
}

#[test]
fn integer_arguments_floor_fractional_values() {
    assert_eq!(
        value("array::length(value: array::filled(count: 3.9, value: 0))").unwrap(),
        Value::Integer(3)
    );
    assert_eq!(
        value("math::random(seed: 3.9, index: 2.9)").unwrap(),
        value("math::random(seed: 3, index: 2)").unwrap()
    );
    assert_eq!(
        value("math::noise(x: 0.1, seed: 3.9)").unwrap(),
        value("math::noise(x: 0.1, seed: 3)").unwrap()
    );
    let s = scene(
        "context 3d render { draw::point_cloud(count: 3.9) draw::grid(columns: 2.9, rows: 3.9) }",
    )
    .unwrap();
    assert_eq!(s.point_clouds[0].count, 3);
    assert_eq!(s.point_clouds[1].grid, Some((2, 3)));
    for v in [
        Value::Float(f64::NAN),
        Value::Float(f64::INFINITY),
        Value::Float(9223372036854775808.0),
        Value::Boolean(true),
    ] {
        assert!(v.floor_integer("test").is_err());
    }
    assert_eq!(Value::Float(-0.2).floor_integer("test").unwrap(), -1);
    assert_eq!(
        Value::Integer(i64::MAX).floor_integer("test").unwrap(),
        i64::MAX
    );
    assert!(value("array::filled(count: -0.2, value: 0)").is_err());
}
